example-app/src/main/java/com/adyen/checkout/example/ui/main/MainViewModel.kt

```diff
@@ -0,0 +1,115 @@
+    val navigateTo: Flow<MainNavigation> = _navigateTo
+
+    fun onComponentEntryClick(entry: ComponentItem.Entry) {
+        when (entry) {
+            ComponentItem.Entry.Card -> _navigateTo.tryEmit(MainNavigation.Card)
+            ComponentItem.Entry.DropIn -> startDropInFlow()
+            ComponentItem.Entry.DropInWithSession -> startSessionDropInFlow()
+        }
+    }
+
+    private fun startDropInFlow() {
+        viewModelScope.launch {
+            _viewState.emit(MainViewState.Loading)
+
+            val paymentMethods = getPaymentMethods()
+            if (paymentMethods != null) {
+                _viewState.emit(MainViewState.Result(ComponentItemProvider.getComponentItems()))
+                _navigateTo.emit(MainNavigation.DropIn(paymentMethods))
+            } else {
+                _viewState.emit(MainViewState.Error("Something went wrong while fetching payment methods"))
+            }
+        }
+    }
+
+    private fun startSessionDropInFlow() {
+        viewModelScope.launch {
+            _viewState.emit(MainViewState.Loading)
+
+            val paymentMethodsDef = async { getPaymentMethods() }
+            val sessionDef = async { getSession() }
+
+            val paymentMethods = paymentMethodsDef.await()
+            val session = sessionDef.await()
+
+            if (paymentMethods != null && session != null) {
+                _viewState.emit(MainViewState.Result(ComponentItemProvider.getComponentItems()))
+                _navigateTo.emit(MainNavigation.DropInWithSession(session, paymentMethods))
+            } else {
+                _viewState.emit(MainViewState.Error("Something went wrong while starting session"))
+            }
+        }
+    }
+
+    private suspend fun getPaymentMethods() = paymentsRepository.getPaymentMethods(
+        getPaymentMethodRequest(
+            merchantAccount = keyValueStorage.getMerchantAccount(),
+            shopperReference = keyValueStorage.getShopperReference(),
+            amount = keyValueStorage.getAmount(),
+            countryCode = keyValueStorage.getCountry(),
+            shopperLocale = keyValueStorage.getShopperLocale(),
+            splitCardFundingSources = keyValueStorage.isSplitCardFundingSources()
+        )
+    )
+
+    private suspend fun getSession() = paymentsRepository.getSessionAsync(
+        getSessionRequest(
+            merchantAccount = keyValueStorage.getMerchantAccount(),
+            shopperReference = keyValueStorage.getShopperReference(),
+            amount = keyValueStorage.getAmount(),
+            countryCode = keyValueStorage.getCountry(),
+            shopperLocale = keyValueStorage.getShopperLocale(),
+            splitCardFundingSources = keyValueStorage.isSplitCardFundingSources(),
+            isExecuteThreeD = keyValueStorage.isExecuteThreeD(),
+            isThreeds2Enabled = keyValueStorage.isThreeds2Enable(),
+            // RedirectComponent.getReturnUrl(applicationContext) should be used here
```

Comment:
Will we solve this later or in this PR?

Reply:
Hmm let me try it in this one. If it gets too big I will open another PR

Reply:
Turns out to be quite easy with `SavedStateHandle`, so also did it for `CardActivity`

Reply:
I wanted to use assisted injection (kind of similar to what you're suggesting), but this is the better approach according to this thread https://github.com/google/dagger/issues/2287